nf_core/download.py

```diff
@@ -298,25 +440,87 @@ def prompt_use_singularity_cachedir(self):
-                "\nIf you are working on the same system where you will run Nextflow, you can leave the downloaded images in the "
-                "[blue not bold]$NXF_SINGULARITY_CACHEDIR[/] folder, Nextflow will automatically find them. "
+                "\nIf you are working on the same system where you will run Nextflow, you can amend the downloaded images to the ones in the"
+                "[blue not bold]$NXF_SINGULARITY_CACHEDIR[/] folder, Nextflow will automatically find them."
                 "However if you will transfer the downloaded files to a different system then they should be copied to the target folder."
             )
-            self.singularity_cache_only = rich.prompt.Confirm.ask(
-                "[blue bold]?[/] [bold]Copy singularity images from [blue not bold]$NXF_SINGULARITY_CACHEDIR[/] to the target folder?[/]"
-            )
+            self.singularity_cache = questionary.select(
+                "[blue bold]?[/] [bold]Copy singularity images from [blue not bold]$NXF_SINGULARITY_CACHEDIR[/] to the target folder or amend new images to the cache?[/]",
+                choices=["amend", "copy"],
+                style=nf_core.utils.nfcore_question_style,
+            ).unsafe_ask()
 
-        # Sanity check, for when passed as a cli flag
-        if self.singularity_cache_only and self.container != "singularity":
-            raise AssertionError("Command has '--singularity-cache-only' set, but '--container' is not 'singularity'")
+    def prompt_singularity_cachedir_remote(self, retry):
+        """Prompt about the index of a remote $NXF_SINGULARITY_CACHEDIR"""
+        if (
+            self.container == "singularity"
+            and self.singularity_cache == "remote"
+            and self.singularity_cache_index is None
+            and stderr.is_interactive  # Use rich auto-detection of interactive shells
+        ):
+            # Prompt user for a file listing the contents of the remote cache directory
+            cachedir_index = None
+            while cachedir_index is None:
+                prompt_cachedir_index = questionary.path(
+                    "Specify a list of the remote images already present in the remote system :",
+                    file_filter="*.txt",
+                    style=nf_core.utils.nfcore_question_style,
+                ).unsafe_ask()
+                cachedir_index = os.path.abspath(os.path.expanduser(prompt_cachedir_index))
+                if prompt_cachedir_index == "":
+                    log.error("Will disregard contents of a remote [blue]$NXF_SINGULARITY_CACHEDIR[/]")
+                    self.singularity_cache_index = None
+                    self.singularity_cache = "copy"
+                elif not os.access(cachedir_index, os.R_OK):
+                    log.error(f"'{cachedir_index}' is not a valid, readable file.")
+                    cachedir_index = None
+            if cachedir_index:
+                self.singularity_cache_index = cachedir_index
```

Comment:
if we have an index, shouldn't this also invoke `read_remote_containers()`

Reply:
Yes, you are right, this is the cleaner solution that also avoids the `retry` complexity.


At first, I invoked `read_remote_containers()` at the top-level in `download_workflow()` separately, because I didn't want to pack non-prompt functionality into a function with `prompt_` in its name, but this way avoids cluttering the code.